### PR TITLE
Add RemainderIs operator

### PR DESCRIPTION
**Pull request by mmoskal**
@hudson-ai - take a look, maybe you have some more ideas for tests? generally appears to work.

I suspect that checking intersection of ReminderIs() with some number ranges might be somewhat expensive, but I guess we'll see on real testcases

see https://github.com/guidance-ai/llguidance/pull/79

### CODE REVIEW

**Comment by hudson-ai** [comment]
Not a super helpful comment, but it should be `Remainder`, not `Reminder`, no? :)

Orthogonal to that, help me wrap my head around something here...
Clearly, we can keep track of the remainder of the digits that have already been written, and we just say that the next digit is nullable iff said remainder is zero. That feels like a clear and simple and "obviously correct" algorithm.

That being said, I can't quite seem to get here following the definition of derivatives!
Let $R(a, k)$ be the regex matching any string that has a value that is $k \pmod{a}$.

Then $\text{deriv}(R(a,k), d) = \lbrace s \mid d\cdot s \in R(a,k)\rbrace$.

We have that $\text{val}(d\cdot s) \equiv \text{val}(d)\cdot 10^{n} + \text{val}(s)$ where $n = \text{len}(s)$.

This seems to imply that

$$
\text{deriv}(R(a,k), d) = \bigcup_{n ≥ 0} \lbrace s \mid \text{len}(s) = n, s \in R(a,k - \text{val}(d)\cdot 10^{n} \pmod{n})\rbrace\.
$$

This is definitely not the desired result.

File: src/bytecompress.rs [hunk 1]
```diff
@@ -103,6 +104,14 @@ impl ByteCompressor {
                 Expr::ByteSet(bs) => {
                     self.bytesets.push(bs.to_vec());
                 }
+                Expr::ReminderIs(_, _) => {
+                    for b in exprset.digits {
+                        if self.mapping[b as usize] == INVALID_MAPPING {
```
Comment (mmoskal) [inline src/bytecompress.rs:109]: Need additional check in byte case!

**Comment by hudson-ai** [comment]
The logic seems correct, although I want to note that as written, we can't get regexes that match $\lbrace s | \text{val}(s) = r \pmod{d}\rbrace$ for $r\neq0$. This is probably super minor, as we likely don't need this to work for $r\neq0$. But adding this will require one extra piece of book-keeping.

In that vein, I worked out a "proper" proof. Been bugging me.

Let $R(d, r, c)$ be the regex matching any string $s$ such that

$$
\text{val}(s) + 10^{|s|}c \equiv r \pmod{d}\,.
$$

By the definition of the regex derivative,

$$
\text{deriv}(R(d,r,c), x) = \lbrace s \mid x\cdot s \in R(d,r,c)\rbrace
$$

We have that $x\cdot s \in R(d,r,c) \iff \text{val}(x\cdot s) + 10^{|s|+1}c$. Furthermore, we have that $\text{val}(x\cdot s) = s + 10^{|s|}\text{val}(x)$.

Therefore, $\text{val}(s) + 10^{|s|}\text{val}(x) + 10^{|s+1|}c \equiv r \pmod{d}$.

Rearranging,

$$
\text{val}(s) + 10^{|s|}(10c + \text{val}(x)) \equiv r \pmod{d}\,,
$$

so

$$
\text{deriv}(R(d,r,c), x) = R(d, r, 10c + \text{val}(x))\,.
$$

Finally, observe that $R(d,r,0)$ matches all strings $s$ such that

$$
\text{val}(s) \equiv r \pmod{d}\,,
$$

as needed.

In terms of an implementation, we just need to say that we're nullable iff `r == c`.

**Comment by hudson-ai** [comment]
I also think we can extend this to decimal $d$ with a little extra bookkeeping. Will mull that over (not essential for this PR)

**Pull request merged**